scripts/babel-relay-plugin/src/RelayQLTransformer.js

```diff
@@ -270,13 +270,6 @@ class RelayQLTransformer {
           'graphql/validation/rules/VariablesInAllowedPosition'
         ).VariablesInAllowedPosition,
       ];
-      if (!isMutation) {
-        rules.push(
-          require(
-            'graphql/validation/rules/ProvidedNonNullArguments'
-          ).ProvidedNonNullArguments
-        );
-      }
```

Comment:
Rewrote the validator to only skip the `ProvidedNonNullArguments` if the top-level definition is a `fragment` with a `relay(pattern: true)` directive.

This doesn't support inline fragments but in a fat query I believe you'd only be using pattern on the top-level fragment.  Supporting inline fragments would require overriding GraphQL's `ProvidedNonNullArguments` rule.